src/vm.rs

```diff
@@ -608,6 +624,155 @@ impl<T: GuestStagePageTable> Vm<T, VmStateFinalized> {
+        }
+    }
+
+    fn handle_pmu_msg(
+        &self,
+        pmu_func: PmuFunction,
+        active_vcpu: &mut ActiveVmCpu<T>,
+    ) -> EcallResult<u64> {
+        use PmuFunction::*;
+        fn get_num_counters() -> EcallResult<u64> {
+            let pmu_info = PmuInfo::get()?;
+            Ok(pmu_info.get_num_counters())
+        }
+
+        fn get_counter_info(counter_index: u64) -> EcallResult<u64> {
+            let pmu_info = PmuInfo::get()?;
+            let info = pmu_info.get_counter_info(LogicalCounterIndex(counter_index))?;
+            Ok(info.raw())
+        }
+
+        fn start_counters<T: GuestStagePageTable>(
+            counter_index: u64,
+            counter_mask: u64,
+            start_flags: PmuCounterStartFlags,
+            initial_value: u64,
+            active_vcpu: &mut ActiveVmCpu<T>,
+        ) -> EcallResult<u64> {
+            let pmu_info = PmuInfo::get()?;
+            let counter_index = LogicalCounterIndex(counter_index);
+            let counter_mask = LogicalCounterMask(counter_mask);
+            let (platform_counter_index, platform_counter_mask) =
+                pmu_info.get_platform_counter_index_and_mask(counter_index, counter_mask)?;
+            active_vcpu.is_counter_range_startable(counter_index, counter_mask)?;
+            let result = pmu::start_counters(
+                platform_counter_index.0,
+                platform_counter_mask.0,
+                start_flags,
+                initial_value,
+            );
+            // Special case "already started" to handle counters that are autostarted following configuration.
+            // Examples of such counters include the legacy timer and insret.
+            if let Err(e) = result && !matches!(e, SbiError::AlreadyStarted) {
+                return Err(EcallError::from(e));
+            }
+            active_vcpu.update_started_counters(counter_index, counter_mask);
+            Ok(0)
+        }
+
+        fn stop_counters<T: GuestStagePageTable>(
+            counter_index: u64,
+            counter_mask: u64,
+            stop_flags: PmuCounterStopFlags,
+            active_vcpu: &mut ActiveVmCpu<T>,
+        ) -> EcallResult<u64> {
+            let pmu_info = PmuInfo::get()?;
+            let counter_index = LogicalCounterIndex(counter_index);
+            let counter_mask = LogicalCounterMask(counter_mask);
+            let (platform_counter_index, platform_counter_mask) =
+                pmu_info.get_platform_counter_index_and_mask(counter_index, counter_mask)?;
+            active_vcpu.is_counter_range_stoppabble(counter_index, counter_mask)?;
+            let result = pmu::stop_counters(
+                platform_counter_index.0,
+                platform_counter_mask.0,
+                stop_flags,
+            );
+
+            // Special case "already stopped" to handle counters can be reset following a stop
+            if let Err(e) = result && !matches!(e, SbiError::AlreadyStopped) {
```

Comment:
same comment as above. AlreadyStopped error should only be ignore if the RESET is set. Otherwise, it should be propagated to VM.

Reply:
Good point -- I am pretty sure I was doing it before in the when update_stopped_counters() was returning an error, but looks like it got lost.